pulp_rpm/tests/functional/api/test_sync.py

```diff
@@ -75,1390 +66,988 @@
-        self.assertEqual(len(task.created_resources), 1)
-        self.assertEqual(publications_api.list().count, 0)
+    with pytest.raises(ApiException):
+        rpm_repository_api.sync(repository.pulp_href, repository_sync_data)
 
-        # 3. create another remote and re-sync
-        body = gen_rpm_remote(url=RPM_SIGNED_FIXTURE_URL)
-        remote = remote_api.create(body)
-        self.addCleanup(remote_api.delete, remote.pulp_href)
 
-        repository_sync_data = RpmRepositorySyncURL(
-            remote=remote.pulp_href, sync_policy=sync_policy
+@pytest.mark.parallel
+def test_sync_skip_srpm_fails_mirror_complete(init_and_sync):
+    """Test that sync is rejected if both skip_types and mirror_complete are configured."""
+    with pytest.raises(ApiException):
+        init_and_sync(
+            url=SRPM_UNSIGNED_FIXTURE_URL, skip_types=["srpm"], sync_policy="mirror_complete"
         )
-        sync_response = repo_api.sync(repo.pulp_href, repository_sync_data)
-        task = monitor_task(sync_response.task)
-
-        # 4. check that one publication was created w/ no repository versions, and only
-        # the new content is present
-        repo = repo_api.read(repo.pulp_href)
-        self.assertDictEqual(RPM_FIXTURE_SUMMARY, get_content_summary(repo.to_dict()))
-        if sync_policy == "mirror_complete":
-            self.assertEqual(publications_api.list().count, 1)
-            self.assertEqual(len(task.created_resources), 2)
-        else:
-            self.assertEqual(len(task.created_resources), 1)
+
+
+@pytest.mark.parallel
+def test_core_metadata(init_and_sync, rpm_package_api):
+    """Test that the metadata returned by the Pulp API post-sync matches what we expect.
+
+    Do the following:
+
+    1. Sync a repo.
+    2. Query package metadata from the API.
+    3. Match it against the metadata that we expect to be there.
+    """
+    repository, _ = init_and_sync(url=RPM_COMPLEX_FIXTURE_URL, policy="on_demand")
+
+    package = rpm_package_api.list(
+        name=RPM_COMPLEX_PACKAGE_DATA["name"], repository_version=repository.latest_version_href
+    ).results[0]
+    package = package.to_dict()
+    # delete pulp-specific metadata
+    package.pop("pulp_href")
+    package.pop("pulp_created")
+
+    # sort file and changelog metadata
+    package["changelogs"].sort(reverse=True)
+    for metadata in [package, RPM_COMPLEX_PACKAGE_DATA]:
+        # the list-of-lists can't be sorted easily so we produce a string representation
+        files = []
+        for f in metadata["files"]:
+            files.append(
+                "{basename}{filename} type={type}".format(
+                    basename=f[1], filename=f[2], type=f[0] or "file"
+                )
+            )
+        metadata["files"] = sorted(files)
+
+    assert all(
+        value == RPM_COMPLEX_PACKAGE_DATA[key]
+        for key, value in package.items()
+        if key != "time_file"
+    )
```

Comment:
So the reason for using dictdiffer was that it printed a more readable representation of what was different, if there was a difference.  We don't have to use that exact method but it would be nice to see why the test failed if it does fail.

It's not worth spending a lot of time since it's passing so I'm OK with leaving it as is currently, just keep in mind the general principle that when it fails it should be obvious why it failed so that it doesn't require a lot of deep debugging.

Reply:
If you want to drop `dictdiffer` though make sure it's also removed from the test requirements

Reply:
A difference between values should be displayed by pytest's nature. Do you think these outputs are not satisfactory: https://docs.pytest.org/en/7.1.x/example/reportingdemo.html#tbreportdemo?

Regarding `dictdiffer`, it is still used in test_publish. Once we refactor that, I am eager to remove the requirement.

Reply:
At least with the old `assertDictEquals` functions it would heavily truncate the results to the point where it was kind of useless.  If pytest doesn't do that then :+1:

Reply:
Okay, after some consideration, I have checked the error outputs and `dictdiffer` is much more "verbose" than the default's pytest utilities. I am reverting the `dictdiffer` removal.